docs/dev/shimplify.md

```diff
@@ -0,0 +1,236 @@
+you can undo with
+
+```bash
+git restore --staged sql-plugin tests
+git restore sql-plugin tests
+```
+
+and by reviewing and removing the new directories with
+
+```bash
+git clean -f -d --dry-run
+```
+
+### Partial Conversion
+
+It is not expected to be really necessary but it is possible to convert a subset of the shims
+
+* Either by adding -Dshimplify.shims=buildver1,buildver2,... to the commands above
+* Or by specifying a list of directories you would like to delete to have a simpler directory
+-Dshimplify.dirs=311until340-non330db,320until330-noncdh
+
+The latter is just a minor twist on the former. Instead of having an explicit list of shims, it
+first computes the list of all `buildver` values using provided directories. After this *all* the
+files for the shims, not just under specified directories are converted.
+
+In both cases, the conversion does not leave the rest of the shims totally unaffected when
+there are common files with a specified shim. However, it guarantees to leave the previous dedicated
+files under `src/(main|test)/${buildver}` in place for shims outside the list. This is useful when
+developers of a certain shim would like to continue working on it without adapting the new method.
+However, for the simplicity of future refactoring the full transition is preferred.
+
+### Evolving shims without automatic conversion
+
+Suppose a bulk-conversion of existing shims is not an option whereas the next shimming issue
+requires difficult refactoring of version ranges with adding more directories with exceptions.
+Now it can be resolved easily by placing just the affected files to owner shim directories and
+adding shim JSON lines comments by hand.
+
+## Adding a new shim
+
+Shimplify can clone an existing shim based as a basis of the new shim. For example when adding
+support for a new [maintenance][5] version of Spark, say 3.2.4, it's expected to be similar to 3.2.3.
+
+If just 3.2.3 or all shims after the full transition have already been converted you can execute
+
+```bash
+mvn generate-sources antrun:run@shimplify=shim-sources -Dshimplify=true \
+    -Dshimplify.move=true -Dshimplify.overwrite=true \
+    -Dshimplify.add.shim=324 -Dshimplify.add.base=323
+```
+
+to clone 323 as 324. This will add `{"spark": "324"}` to every shared file constituting the 323
+shim. Moreover, it will create
+
+* a copy of dedicated 323 files with spark323 under spark324 shim
+directory
+* substitute spark324 for spark323 in the package name and path,
+* and modify the comment from `{"spark": "323"}` to `{"spark": "324"}`
+
+Review the new repo state, e.g., using `git grep '{"spark": "324"}'`.
+Besides having to add the `release324` profile to various pom.xml as before, this alone
+is likely to be insufficient to complete the work on 324. It is expected you will need to
+work on resolving potential compilation failures manually.
+
+## Deleting a Shim
```

Comment:
Probably makes sense to have a follow-on to pass a flag to drop support of a shim from command-line and the shimplify.py should take care of all of this.